build-data-capturing-gradle-samples/capture-quality-check-issues/gradle-quality-check-issues.gradle

```diff
@@ -10,100 +13,143 @@ if (!buildScanApi) {
-            it.File.collect {
-                String filePath = task.project.rootProject.relativePath(sourceDirectory + (it.@name.text() as String))
-                it.Violation.collect { "${filePath}:${it.@lineNumber} \u2192 ${it.Message.text()}" }
-            }.flatten()
-        }.flatten() as List<String>
-    } else if (ReportingTask.FINDBUGS.isTask(task)) {
-        valueName = 'Verification FindBugs'
-        errors = report.BugInstance.collect { bugInstance ->
-            String type = bugInstance.@type.text()
-            def error = bugInstance.breadthFirst().findAll { it.name() == 'SourceLine' }.sort(false) { it.parent().name() as SpotBugsParent }.first()
-            String startLine = error.@start.text()
-            String endLine = error.@end.text()
-            String lineNumber = endLine == startLine ? startLine : "$startLine-$endLine"
-            def absoluteFilePath = bugInstance.parent().Project.SrcDir.collect { it.text() }.find { it.endsWith(error.@sourcepath.text()) }
-            String relativePath = task.project.rootProject.relativePath(absoluteFilePath)
-            "${relativePath}:${lineNumber} \u2192 ${type}"
-        }.flatten() as List<String>
-    } else if (ReportingTask.SPOTBUGS.isTask(task)) {
-        valueName = 'Verification SpotBugs'
-        errors = report.BugInstance.collect { bugInstance ->
-            String type = bugInstance.@type.text()
-            def error = bugInstance.breadthFirst().findAll { it.name() == 'SourceLine' }.sort(false) { it.parent().name() as SpotBugsParent }.first()
-            String startLine = error.@start.text()
-            String endLine = error.@end.text()
-            String lineNumber = endLine == startLine ? startLine : "$startLine-$endLine"
-            def absoluteFilePath = bugInstance.parent().Project.SrcDir.collect { it.text() }.find { it.endsWith(error.@sourcepath.text()) }
-            String relativePath = task.project.rootProject.relativePath(absoluteFilePath)
-            "${relativePath}:${lineNumber} \u2192 ${type}"
-        }.flatten() as List<String>
-    }
-
-    errors.each { buildScan.value valueName, it }
 }
 
-static String appendIfMissing(String str, String suffix) {
-    str.endsWith(suffix) ? str : str + suffix
+// check if reporting tasks are present in the build
+def hasReportingTask = false
+gradle.taskGraph.whenReady { taskGraph ->
+    if(taskGraph.allTasks.findAll {
+        it.class.name.startsWith('org.gradle.api.plugins.quality.Checkstyle')
+            || it.class.name.startsWith('org.gradle.api.plugins.quality.CodeNarc')
+            || it.class.name.startsWith('org.gradle.api.plugins.quality.FindBugs')
+            || it.class.name.startsWith('com.github.spotbugs.SpotBugsTask')
+    }.size() > 0) {
+        hasReportingTask = true
+    }
 }
 
-enum SpotBugsParent {
-
-    BugInstance, Method, Class
+// process results on build completion
+buildScan.buildFinished { def result ->
```

Comment:
what does the `def result ->` here do?

Reply:
Will vanished eventually as `buildScan.buildFinished` is not required anymore